### PR TITLE
feat: add custom key to projects table, backfilling based on current project name, and API support

### PR DESCRIPTION
**Pull request by corban-beaird**
## Ticket
ET-97
ET-101
ET-102
ET-164

## Description
Introduce custom project key (similar to JIRA's project key), to be able to refer to projects. This adds the column to the DB & backfills existing projects based on their existing project name.

This also includes allowing the key to be specified in the `CreateProject` endpoint (PostProject). 

## Test Plan
- Automated Integration Testing 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Commentary
Work will need to be done to allow updating project keys of existing projects (that is being handled in ET-98). This will likely include updating the logic for how we determine uniqueness, since the requirement will need to ensure no past project keys had been in use.

### CODE REVIEW

File: master/internal/api_project_intg_test.go [hunk 1]
```diff
@@ -371,3 +371,67 @@ func TestGetProjectByActivity(t *testing.T) {
...
+	err = db.Bun().NewSelect().
+		Column("key").
+		Table("projects").
+		Where("id = ?", resp.Project.Id).
+		Scan(ctx, &resp.Project.Key)
+	require.NoError(t, err)
+	require.Equal(t, projectKey, resp.Project.Key)
+}
+
+func TestCreateProjectWithDuplicateProjectKey(t *testing.T) {
```
Comment (AmanuelAaron) [inline master/internal/api_project_intg_test.go:420]: we should add a test case for creating a project without a project key that results in a duplicate project key. In other words they have the same first 3 chars

File: master/internal/api_project.go [hunk 2]
```diff
@@ -523,9 +524,29 @@ func (a *apiServer) PostProject(
...
-		req.WorkspaceId, curUser.ID)
+		req.WorkspaceId, curUser.ID, projectKey)
+
+	if err != nil && strings.Contains(err.Error(), db.CodeUniqueViolation) {
+		if strings.Contains(err.Error(), "projects_key_key") {
+			return nil,
+				status.Errorf(codes.AlreadyExists, "project with key %s already exists", projectKey)
+		} else if strings.Contains(err.Error(), "projects_name_key") {
+			return nil,
+				status.Errorf(codes.AlreadyExists, "project with name %s already exists", req.Name)
```
Comment (AmanuelAaron) [inline master/internal/api_project.go:547]: Any reason why we can have 2 projects with the same name?

Comment (corban-beaird) [reply master/internal/api_project.go:547]: My motivation for this was to return a 403 (already exists) status instead of returning an 500 since this wasn't an internal error rather an issue with the patch request.

File: master/internal/api_project_intg_test.go (hunk 1 above)
Comment (corban-beaird) [reply master/internal/api_project_intg_test.go:420]: Good call! I can confirm that behavior works based on manual testing, but that's a great test to include!

File: master/internal/api_project.go (hunk 2 above)
Comment (corban-beaird) [reply master/internal/api_project.go]: However, I'll remove it, this was more of a nit of my part for how we communicate errors. I think making sure that we are returning appropriate response codes is something we could improve.

File: master/internal/api_project.go [hunk 3]
```diff
@@ -535,9 +536,26 @@ func (a *apiServer) PostProject(
...
+			return nil, fmt.Errorf("error generating project key: %w", err)
+		}
+	} else {
+		projectKey = *req.Key
+	}
+
 	p := &projectv1.Project{}
 	err = a.m.db.QueryProto("insert_project", p, req.Name, req.Description,
-		req.WorkspaceId, curUser.ID)
+		req.WorkspaceId, curUser.ID, projectKey)
```
Comment (AmanuelAaron) [inline master/internal/api_project.go:551]: Worried about a race condition here where two projects with the same first 3 chars are being created at the same time. Both of them would generate the same key. Perhaps we can create the project first without the key, then generates the key filtering out projects created after its own "created_at" time.

**Review by AmanuelAaron — COMMENTED** [review]
Looks good overall.

File: master/internal/api_project.go (hunk 3 above)
Comment (corban-beaird) [reply master/internal/api_project.go:551]: I agree that the race condition here will cause a rough UX, "if no key is specified during project creation, but getting an error about a key conflict" doesn't really make sense.

The suggest approach would require dropping the unique constraint on the key column (which might not be the worst move, since I think that is really just a bandaid on an application level solution for how we handle uniqueness on project keys). I'd also lean away from using the created_at time, since it's susceptible to clock drift/syncs, and could cause some funky flakes. We could piggy back of the strict event ordering work done by streaming updates and leverage the seq value (would do the same thing we'd want to do with a timestamp, without the risk of clock drift).

File: master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    -- Take the first 3 characters of the input string
+    prefix := LEFT(input_string, 3);
+	
+   	execute format('SELECT COUNT(*)+1 FROM projects WHERE key ILIKE $1') 
+   		into count_value
+   		using prefix || '%';
+   	count_suffix := CAST(count_value as text);
+
+    -- Concatenate prefix and count suffix
+    RETURN lower(prefix || count_suffix);
```
Comment (gt2345) [inline master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: I'm not too sure about this way of generating project keys. Imagine that we have a project with key `aa12`, then we have a project with name `aa1`, running this function would run into duplicated keys.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: Yea, you have a good point, I don't think the case you brought up is an immediate issue, assuming this function is only used to backfill keys for existing projects, but it will be a problem if it gets used to generate keys for new projects when there are existing keys. 

If this is only being used to backfill:
- if we have a project key of `aa12`, that would mean that:
    - this is either the 12 instance of the prefix `aa` (assuming the name of the project is `aa`)
    - or the 2nd instance of the prefix `aa1` (assuming the name of the project is `aa1`)
- so if we have a project with the prefix `aa1`
    - the count would be 3 since we only care about the keys that start with `aa1` 
    - since there is only going to be 2 or 3 other cases where that string appears as a key
        - `aa11` and `aa12` (and possibly `aa1`)

That being said, I do see a case where:
- project name: `aa1...`
- resulting key: `aa11`
- If there are 10 other projects with the name `aa` (can happen with enough workspaces)
- then we'd end up with:
    - `aa1` -> `aa11`
    - `aa` -> `aa2`
    - `aa` -> `aa3`
    - `aa` -> `aa4`
    - ... 
    - `aa` -> `aa10`
    - `aa` -> `aa11` << error on duplicate key

Now I don't think this case is as likely to happen since it's requires the project name to only be 2 characters, there to be identical 2 character projects across 10 different workspaces; but it does demonstrate it's possible.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: If we sort the way these things are handled to avoid cases like this, so instead of handling any project with `aa1...` first, we'd instead handle all the `aa` projects first, and then that would result with the  `aa1...` project being assigned the key `aa12` instead of `aa11`.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: That doesn't put us in the clear for using this style of approach for generating keys for new projects since we can't order how those will be processed or how they had been handled previously.

Comment (gt2345) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: For new project, since we allow passing key when creating, it would by alright to just reject duplicated key as long as we have a clear error message

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: The catch is that we don't require the key to be specified, and I'm worried that if we return an error about a duplicate key that the user didn't provide it would lead to some confusion. 

The more I'm thinking about this as well, this approach also exposes potentially sensitive information about what project names (or at least the first 3 characters) exist in the system that the user making the request might not have access to. But that might be a none issue.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: The reason why I think it could be a none issue is because having incremental project keys also runs up against this, since you'll be able to tell how many projects had historically existing (they might not all exist still if they got deleted, but at some point they had).

Comment (gt2345) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: Suppose we have keys `a1 a2 a3` in the system, then we delete the row associated with `a1`, then a newly created project with name `a` would have the generated key to be `a3`?

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: That is also an issue with this approach if used for anything outside of backfilling. If instead of counting we parse the number at the end of the key, and get the max from the set, then we could avoid that problem, but then we get into a tough spot since the generated keys will never re-use earlier values. 

Even worse, someone could specify a key of `a99999` and if we don't have way to cycle back to free keys we'll be stuck.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: We'd be in a position where we'd have to search for the free key, we could do a binary search by comparing the max & count in each subset; but this all seems like over engineering.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: If we really want something like this, I could imagine instead of basing it off an incremental value, keep the 1-3 character prefix from the project name + 2-4 random characters (w/ a retry). 

That avoids the RBAC concern outside of key collisions + avoids never re-using freed keys. This does mean that the key generation performance is probabilistic and degrades as we start hitting the upper limits of a given prefix. 

If it wasn't a product requirement to have the generated keys be based on the names, either assigning this to be 5 random alphanumeric character (or just be equivalent to the project id by default), we could avoid this problem entirely.

Comment (corban-beaird) [reply master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql:17]: Plus using the random character suffix means we don't have to enforce and order on the update & can use the same function to generate keys outside of backfill

File: master/static/migrations/20240508154139_add-custom-project-key.tx.up.sql [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+END;
+$$ LANGUAGE plpgsql;
+
+ALTER TABLE projects ADD COLUMN IF NOT EXISTS key VARCHAR(5) UNIQUE;
+
+UPDATE 
+    projects 
+SET 
+    key = function_generate_project_key(5, 3, name)
+WHERE key IS NULL;
```
Comment (gt2345) [inline master/static/migrations/20240508154139_add-custom-project-key.tx.up.sql:33]: Should we have a migration down file?

File: master/internal/project/postgres_project.go [hunk 6]
```diff
@@ -50,3 +66,61 @@ func ProjectIDByName(ctx context.Context, workspaceID int, projectName string) (
 	}
 	return &pID, nil
 }
+
+// GenerateProjectKey generates a unique project key for a project based on its name.
+func generateProjectKey(ctx context.Context, tx bun.Tx, projectName string) (string, error) {
```
Comment (gt2345) [inline master/internal/project/postgres_project.go:71]: Why do we switch from using sql function to using go function?

**Review by gt2345 — COMMENTED** [review]
Do we plan to allow changing project key after creation? My concern is that, it's possible that the generated key is an annoying/racist word, and the user might want to change that

File: master/static/migrations/20240508154139_add-custom-project-key.tx.up.sql (hunk 5 above)
Comment (corban-beaird) [reply master/static/migrations/20240508154139_add-custom-project-key.tx.up.sql:33]: Nope, down migrations got removed recently 🥳

File: master/internal/api_project.go [hunk 7]
```diff
@@ -520,6 +526,19 @@ func (a *apiServer) getProjectNumericMetricsRange(
...
+	case len(key) < 1:
+		return errors.New("project key cannot be empty")
+	case !regexp.MustCompile(ProjectKeyRegex).MatchString(key):
+		return errors.Errorf("project key can only contain alphanumeric characters")
+	default:
+		return nil
+	}
+}
+
 func (a *apiServer) PostProject(
```
Comment (AmanuelAaron) [inline master/internal/api_project.go:542]: Do we want to add the ability to change the project key. Or do we want to add that in a separate ticket?

File: master/internal/api_project_intg_test.go
Comment (AmanuelAaron) [inline master/internal/api_project_intg_test.go:1]: Could we add a test case for when someone adds with a key that ends in a number like "TEST99". Additionally in that case what would happen if we try to add a another project with that same key

Comment (corban-beaird) [reply master/internal/api_project_intg_test.go:1]: I changed the approach that we're using to generate the project keys to a random set alphanumeric characters, this way we avoid issues wheres someone has number at the postfix.

**Review by AmanuelAaron — APPROVED** [review]

File: master/internal/project/postgres_project.go [hunk 8]
```diff
@@ -50,3 +136,285 @@ func ProjectIDByName(ctx context.Context, workspaceID int, projectName string) (
...
+}
+
+// InsertProject inserts a new project into the database.
+func InsertProject(
+	ctx context.Context,
+	p *model.Project,
+	requestedKey *string,
+) (err error) {
+RetryLoop:
+	for i := 0; i < MaxRetries; i++ {
```
Comment (gt2345) [inline master/internal/project/postgres_project.go:194]: May I please ask what was the thought behind retrying here given that we already retried in `generateProjectKey`?

File: master/internal/project/postgres_project.go [hunk 9]
```diff
@@ -50,3 +136,285 @@ func ProjectIDByName(ctx context.Context, workspaceID int, projectName string) (
...
+}
+
+// UpdateProjectKey updates the key of a project.
+func UpdateProjectKey(ctx context.Context, projectID int, key string) error {
+	if len(key) > MaxProjectKeyLength {
+		return fmt.Errorf("project key must be at most %d characters", MaxProjectKeyLength)
+	}
+	if projectID == 0 {
+		return fmt.Errorf("invalid project ID")
+	}
```
Comment (gt2345) [inline master/internal/project/postgres_project.go:244]: Can we call `ValidateProjectKey` here? Also, where is this used?

**Review by gt2345 — COMMENTED** [review]
Working as expected when testing on CLI, just a few questions

File: master/internal/project/postgres_project.go (hunk 9 above)
Comment (corban-beaird) [reply master/internal/project/postgres_project.go:244]: Good call out! This approach had been swapped out to follow suit with PatchProject, so this function can be dropped.

File: master/internal/project/postgres_project.go (hunk 8 above)
Comment (corban-beaird) [reply master/internal/project/postgres_project.go:194]: We retry in the `generateProjectKey` case to identify a project key that hasn't been used yet; however, if there are multiple request that come in we'll potentially run into a serializability error if there are concurrent inserts (i.e. two request generate the same project key and attempt to insert at the same time).

**Review by gt2345 — APPROVED** [review]
